### PR TITLE
refactor: x-platform multicall

### PR DESCRIPTION
**Pull request by zzmp**
- Moves multicall into `src/lib`
- Moves block number from the application state into its own jotai atom. This allows it (and multicall, which relies on it) to be used outside of the redux application (ie in widgets)

`useTokenBalances` (in `useQueriedTokenList.ts`) was used as a motivation for this change, so some changes to it and related files (ie `TokenSelect/`) are included as well.

### CODE REVIEW

File: src/lib/components/TokenSelect/useQueriedTokenList.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  const debouncedQuery = useDebounce(query, 200)
+  const filteredTokens = useMemo(
+    () => sortedTokens.filter(createTokenFilterFunction(debouncedQuery)),
+    [debouncedQuery, sortedTokens]
+  )
+
+  // Re-sorts tokens (ie to bump up exact matches)
+  const sortedFilteredTokens = useSortedTokensByQuery(filteredTokens, debouncedQuery)
+
+  // TODO(zzmp): Include native Currency
```
Comment (ianlapham) [inline src/lib/components/TokenSelect/useQueriedTokenList.ts:31]: This is a requirement for merge right? If app will be using this

File: src/state/transactions/updater.tsx [hunk 2]
```diff
@@ -112,11 +112,6 @@ export default function Updater(): null {
                 hash,
                 isL2 ? L2_TXN_DISMISS_MS : DEFAULT_TXN_DISMISS_MS
               )
-
-              // the receipt was fetched before the block, fast forward to that block to trigger balance updates
-              if (receipt.blockNumber > lastBlockNumber) {
```
Comment (ianlapham) [inline src/state/transactions/updater.tsx:117]: Can we keep this logic by just adding a jotai hook to fast forward block number

File: src/lib/hooks/useBlockNumber.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  const block = useBlock()
+  useEffect(() => {
+    setBlock(block)
+  }, [block, setBlock])
+  return null
+}
+
+/** Requires that BlockUpdater be installed in the DOM tree. */
+export default function useBlockNumber() {
+  return useAtomValue(blockAtom)
```
Comment (ianlapham) [inline src/lib/hooks/useBlockNumber.ts:61]: the old code returns undefined if no `chainId` , should we also keep that pattern here?

File: src/state/application/hooks.ts [hunk 4]
```diff
@@ -2,16 +2,9 @@ import { DEFAULT_TXN_DISMISS_MS } from 'constants/misc'
...
 import { useAppDispatch, useAppSelector } from 'state/hooks'
 
-import { useActiveWeb3React } from '../../hooks/web3'
 import { AppState } from '../index'
 import { addPopup, ApplicationModal, PopupContent, removePopup, setOpenModal } from './reducer'
 
-export function useBlockNumber(): number | undefined {
-  const { chainId } = useActiveWeb3React()
-
-  return useAppSelector((state: AppState) => state.application.blockNumber[chainId ?? -1])
```
Comment (ianlapham) [inline src/state/application/hooks.ts:12]: This is that old check for `chainId`

**Review by ianlapham — COMMENTED** [review]
Just a few nits about checking for chainId

File: src/index.tsx [hunk 5]
```diff
@@ -39,6 +40,7 @@ function Updaters() {
       <UserUpdater />
       <ApplicationUpdater />
       <TransactionUpdater />
+      <BlockUpdater />
       <MulticallUpdater />
```
Comment (zzmp) [inline src/index.tsx:44]: Do I need to include a MulticallUpdater in the widget?

File: src/lib/components/TokenSelect/useQueriedTokenList.ts (hunk 1 above)
Comment (zzmp) [reply src/lib/components/TokenSelect/useQueriedTokenList.ts:31]: This is a requirement for launch, but may be deferred to a later PR

File: src/lib/hooks/useBlockNumber.ts (hunk 3 above)
Comment (zzmp) [reply src/lib/hooks/useBlockNumber.ts:61]: Good call. I'll add this back in.

File: src/state/transactions/updater.tsx (hunk 2 above)
Comment (zzmp) [reply src/state/transactions/updater.tsx:117]: Done.

File: src/index.tsx (hunk 5 above)
Comment (zzmp) [reply src/index.tsx:44]: > you have to render it or you won't get any data

https://uniswapteam.slack.com/archives/C023ZR8Q1DE/p1641946622034500?thread_ts=1641946408.032600&cid=C023ZR8Q1DE

**Pull request closed without merging**